src/metrics.jl

```diff
@@ -220,13 +220,42 @@ result_type(dist::UnionMetrics, ::Type{Ta}, ::Type{Tb}, ::Nothing) where {Ta,Tb}
 result_type(dist::UnionMetrics, ::Type{Ta}, ::Type{Tb}, p) where {Ta,Tb} =
     typeof(_evaluate(dist, oneunit(Ta), oneunit(Tb), oneunit(eltype(p))))
 
-Base.@propagate_inbounds function _evaluate(d::UnionMetrics, a::AbstractArray, b::AbstractArray)
+Base.@propagate_inbounds function _evaluate(d::UnionMetrics, a, b)
     _evaluate(d, a, b, parameters(d))
 end
-_evaluate(dist::UnionMetrics, a::Number, b::Number) = _evaluate(dist, a, b, parameters(dist))
 
 # breaks the implementation into eval_start, eval_op, eval_reduce and eval_end
 
+Base.@propagate_inbounds function _evaluate(d::UnionMetrics, a, b, ::Nothing)
```

Comment:
I have tried to that in ef6c646. That commit works, and it seems to distribute the exact same code into inlined helper functions, and yet it introduces patterns like this in `@code_typed euclidean(rand(100), rand(100))`:
```
%119 = Base.add_int(%100, 1)::Int64
│            $(Expr(:loopinfo, Symbol("julia.simdloop"), nothing))::Any
└────        goto #37
45 ┄─ %122 = φ (#37 => %99, #35 => 0.0)::Float64
46 ┄─ %123 = φ (#45 => %122, #34 => 0.0)::Float64
│     %124 = Base.lt_float(%123, 0.0)::Bool
└────        goto #48 if not %124
47 ──        invoke Base.Math.throw_complex_domainerror(:sqrt::Symbol, %123::Float64)::Union{}
└────        $(Expr(:unreachable))::Union{}
48 ┄─ %128 = Base.Math.sqrt_llvm(%123)::Float64
└────        goto #49
49 ──        goto #50
50 ──        goto #98
51 ──        Base.Math.sqrt_llvm(0.0)::Float64
│     %133 = Base.bitcast(UInt64, 1)::UInt64
│     %134 = Base.bitcast(UInt64, 1)::UInt64
│     %135 = Base.sub_int(%133, %134)::UInt64
│     %136 = Base.arraylen(a)::Int64
│     %137 = Base.sle_int(0, %136)::Bool
│     %138 = Base.bitcast(UInt64, %136)::UInt64
│     %139 = Base.ult_int(%135, %138)::Bool
```
I'm not very good at reading this, but it seems like after each loop iteration, it is checking whether it should apply `sqrt`. Also, the `Base.bitcast` calls all completely absent in the current master version. I hope I have done stupid, easily fixable mistake, because I kind of like the design in that commit.

Reply:
I believe one can measure these extra operations in `@btime`, though the slow-down is not dramatic.

Reply:
To be clear, you're comparing the current:
```julia
            @inbounds for (Ia, Ib) in zip(eachindex(a), eachindex(b))
                ai = a[Ia]
                bi = b[Ib]
                s = eval_reduce(d, s, eval_op(d, ai, bi))
            end
```
with the new:
```julia

        @inbounds for (ai, bi) in zip(a, b)
            s = eval_reduce(d, s, eval_op(d, ai, bi))
        end
```

right?

Reply:
No, I'm comparing the `@simd`ed loops paths. You may need to turn off the diff show to see better what the path is in the last commit.

Reply:
OK. I'm not completely sure how to read that either, but `goto #37` seems to indicate that the core of the loop won't reach the `sqrt` call: that will only happen at the end. `@code_llvm` doesn't show any `sqrt` call in the SIMD loop either.

But if you say you can see the difference when benchmarking, maybe reintroduce the old loop. Given that we need to keep the check for the `@simd` path, the deduplication is more limited than I hoped anyway.